File: 800-1000/968. Binary Tree Cameras/solution.js
var minCameraCover = function(root) {
    return helper(root)
};

//recursion
//lights that only cover itself => 1 (can move up to its parent)
//lights that cover itself and its children => 2 (can't move)
//spots covered by its children => 3
let helper = (root) => {
    if (!root) {
        return 0
    } else {
        let res = helper(root.left) + helper(root.right)
        
        if (!root.left && !root.right) root.val = 1
        check(root,root.left)
        check(root,root.right)
        if (root.left && root.left.val === 1 && root.right && root.right.val === 1) res--
        if (root.val === 1) res++
        return res
    }
}


let check = (root,children) => {
    if (children) {
        let childVal = children.val
        let currentVal = root.val

        if (childVal === 3) root.val = 1
        if (currentVal === 3 || childVal === 2) root.val = 3
        if (childVal === 1 || currentVal === 2) root.val = 2
    }
    return
}

File: 800-1000/968. Binary Tree Cameras/solution.cpp
class Solution {
public:
    int minCameraCover(TreeNode* root) {
        queue<TreeNode*> qu;
        qu.push(root);
        int k = 0;
        while (!qu.empty()) {
            auto node = qu.front(); qu.pop();
            node->val = k++;
            if (node->left) qu.push(node->left);
            if (node->right) qu.push(node->right);
        }
        
        vector<vector<int>> dp(k,vector<int>(3,-1));
        return helper(dp,root,2);
    }
    
private:
    int helper(vector<vector<int>>& dp, TreeNode* node, int status) {
        if (!node) return status == 3 ? 1001 : 0;
        
        int val = node->val;
        if (dp[val][status-1]>-1) return dp[val][status-1];
        
        int ans;
        if (status == 1) {
            int dontPut = helper(dp,node->left,2) + helper(dp,node->right,2);
            int putHere = helper(dp,node->left,1) + helper(dp,node->right,1) + 1;
            ans = min(dontPut, putHere);
        } else if (status == 2) {
            int putHere = helper(dp,node->left,1) + helper(dp,node->right,1) + 1;
            int putLeft = helper(dp,node->left,3) + helper(dp,node->right,2);
            int putRight = helper(dp,node->left,2) + helper(dp,node->right,3);
            ans = min(putHere,min(putLeft,putRight));
        } else {
            ans = helper(dp,node->left,1) + helper(dp,node->right,1) + 1;
        }
        
        return dp[val][status-1] = ans;
    }
};

/*
for every node, it can be monitored by itself, its children or its parents

node: the node we're at
status: the three states of a node => whether it has a cam, and whether it is covered
1. it doesn't have a cam but is covered => no need for a cam
2. it doesn't have a cam and is not covered => can put a cam here, or put it at one of its children
3. it does have a cam and is covered => all its children are covered

x(node,status): for every node, given its status when arriving to it, return the minimum amount of cameras required for its represented subtree.
*/
